Enable compression in netty

deepthidevaki opened the pull request:
## Description

* Add two compression algorithms to netty (gzip, snappy)
* Compression can be enabled by configuration (default = disabled)
* When compression is enabled all messages sent between all brokers and gateway will be compressed.

In the spike, we observed that compression increased the throughput when network latency is high. In no latency test, we found that enabling compression has not significant performance impact. 

## Related issues

closes #8486

<!-- Cut-off marker
_All lines under and including the cut-off marker will be removed from the merge commit message_

## Definition of Ready

Please check the items that apply, before requesting a review.

You can find more details about these items in our wiki page about [Pull Requests and Code Reviews](https://github.com/camunda-cloud/zeebe/wiki/Pull-Requests-and-Code-Reviews).

* [x] I've reviewed my own code
* [x] I've written a clear changelist description
* [x] I've narrowly scoped my changes
* [x] I've separated structural from behavioural changes
-->

## Definition of Done

<!-- Please check the items that apply, before merging or (if possible) before requesting a review. -->

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [x] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [x] The impact of the changes is verified by a benchmark 

Documentation: 
* [x] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

deepthidevaki commented:
Throughput Snappy
![image](https://user-images.githubusercontent.com/1997478/147927737-195490e1-c788-4419-9391-44452c5267e5.png)

Throughput Gzip
![image](https://user-images.githubusercontent.com/1997478/147927770-c1f66079-6dbb-4c9f-b49f-b93d1f94b8fe.png)

Compression Snappy:
![image](https://user-images.githubusercontent.com/1997478/147928003-e78871a8-6a2a-43f0-ae2a-2afe1f529bfc.png)

Compression GZip
![image](https://user-images.githubusercontent.com/1997478/147927878-94c39bab-7952-40ea-a57a-20ef4213a8ac.png)

npepinpe commented:
The snappy one seems within normal variation imo, unless that was the max it ever reached...and even then, I would still assume it was fairly similar.

Did you see Ole's last comment about using zstd instead of gzip?

deepthidevaki commented:
> 

Yes. I would suggest to add zstd as a follow up.

oleschoenburg commented on gateway/src/main/java/io/camunda/zeebe/gateway/impl/configuration/CompressionCfg.java:
```diff
@@ -0,0 +1,61 @@
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(enableCompression, algorithm);
```
:upside_down_face: strictly speaking, this isn't the IntelliJ default style: https://github.com/camunda-cloud/zeebe/wiki/Code-Style#equalshashcode-implementations

Same goes for `MembershipCfg`

deepthidevaki replied:
:+1: Other configuration classes uses this format though. I will update the new code to follow the code style.

oleschoenburg reviewed (approved):
Looks great! :+1: 

:wrench: Personally, I wouldn't mind a a different configuration scheme where we only offer `zeebe.broker.cluster.messageCompression = none|gzip|snappy` with none being the default. 

:thought_balloon: : Do we want an end-to-end test for this? It's similar to the new TLS feature where we also don't have an e2e test (I think) so it might be fine but at least to me it's not immediately obvious that `NettyMessagingServiceCompressionTest` ensures this will work (and keep working) for broker to broker and gateway communication.

npepinpe commented:
(just an FYI, we do have a QA test for cluster TLS called `SecureClusteredMessagingIT` :wink:)

deepthidevaki commented on dist/src/main/config/broker.standalone.yaml.template:
```diff
@@ -325,6 +325,20 @@
+        # when the network latency between the brokers is very high (for example when the brokers are deployed in different data centers).
+        # When latency is high, the network bandwidth is severely reduced. Hence enabling compression helps to improve the throughput.
+        # Note: When there is no latency enabling this may have a performance impact.
+        # Note: When this flag is enables, you should also enable compression in standalone gateway configuration.
```
Yes. It is a "must". I will update the doc.
I don't know exactly, but I remember seeing an exception from the decoder.

deepthidevaki commented:
> Looks great! +1
> 
> wrench Personally, I wouldn't mind a different configuration scheme where we only offer `zeebe.broker.cluster.messageCompression = none|gzip|snappy` with none being the default.

I like it. I will change it. 

> 
> thought_balloon : Do we want an end-to-end test for this? It's similar to the new TLS feature where we also don't have an e2e test (I think) so it might be fine but at least to me it's not immediately obvious that `NettyMessagingServiceCompressionTest` ensures this will work (and keep working) for broker to broker and gateway communication.

I can add a integration test that verifies "it works" when compression is enabled. Is that enough? I don't think there is a way to check if messages are actually compressed.

npepinpe commented:
I have a dumb question, since we talked about this with Ole yesterday. The Netty channel pipeline modifies itself - once the protocol handshake is finished, it removes that handler, and adds other handlers at the end of the pipeline. Does that affect the compression? How does the pipeline handle it? Does it mean after the handshake, the new things are added at the end of the pipeline after the compression encoding? :thinking:

deepthidevaki commented:
> I have a dumb question, since we talked about this with Ole yesterday. The Netty channel pipeline modifies itself - once the protocol handshake is finished, it removes that handler, and adds other handlers at the end of the pipeline. Does that affect the compression? How does the pipeline handle it? Does it mean after the handshake, the new things are added at the end of the pipeline after the compression encoding? thinking I guess not, but I don't understand how it works sweat_smile

I don't know what you are referring to. I only know that the handlers are executed in the specific order (one order for inbound and reverse order for outbound).

deepthidevaki commented:
@oleschoenburg I changed the configuration option to `cluster.messageCompression = NONE | GZIP | SNAPPY`. Since there are lot of changes, I squashed the commits and force-pushed them. Could you please review it again?

npepinpe commented:
If you look at the `ClientHandshakeHandlerAdapter` (which is originally added to the pipeline before the compression stuff), in `activateProtocolVersion` it will remove itself, and then add new encoder/decoder and message dispatcher handlers to the pipeline. I'm wondering how this interacts with the compression handlers.

So originally, the pipeline is:

Inbound: `TLS` -> `ClientHandshakeHandlerAdapter` -> `CompressionDecoder`
Outbound: `CompressionEncoder` -> `ClientHandshakeHandlerAdapter` -> `TLS`

Then the pipeline is modified after the initial handshake:

Inbound: `TLS` -> `CompressionDecoder` -> `ProtocolDecoder` -> `MessageDispatcher`
Outbound: `ProtocolEncoder` -> `CompressionEncoder` -> `TLS`

The final pipeline looks good and what I would expect, but the initial one looks weird - and I have a feeling we ended up like that unexpectedly? I always have a hard time wrapping my head around this self modifying pipeline. It might be worth having a follow up issue where the pipeline is fixed but stateful (a bit like the SSL handler, where it knows the handshake was performed so it can now do other stuff). Also, I guess for the initial outbound pipeline, the handshake adapter should be before the compression encoder? I'm not sure why that didn't break anything :man_shrugging:

oleschoenburg reviewed (approved):
:tada: LGTM, thanks for updating the config scheme and writing the e2e test.

oleschoenburg commented:
> The final pipeline looks good and what I would expect, but the initial one looks weird

Hm, I guess that just means that the initial handshake does not use compression, which makes sense. I suppose it'd be also OK to only install the compression handlers after a successful handshake?

npepinpe commented:
It still seems like the end result is unexpected? :smile: I can create a follow up issue to sort it out and make it clearer.

deepthidevaki commented:
@npepinpe Please create a follow up issue. I don't really understand the problem you found. 
I guess I can merge this, as it seems to work? :slightly_smiling_face:

deepthidevaki commented:
bors merge

npepinpe commented:
I think the fact none of us understand it is the real problem :smile:
